### ISSUE TITLE
DATA-1 — Format-agnostic dataset API: `DataBackend` Protocol, registry, XYZ backend, unified statistics

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1569]
**Depends on:** CORE-3 (#1557) · **Blocks:** DATA-2 (#1570), TRN-3 (#1577)

**Prerequisite:** the Phase 0 characterization suite is committed — goldens, fixtures and the single-source tolerance table under `tests/golden/`, characterization tests under `tests/unit/`. References of the form `P0-N` name that work.

**Context:** Legacy has four ingestion paths sharing one *implicit* contract: the in-memory XYZ path eagerly builds the full `list[AtomicData]` (`mace/tools/run_train_utils.py:66-75`), while HDF5 (`mace/data/hdf5_dataset.py`) and LMDB (`mace/data/lmdb_dataset.py`) each rebuild graphs themselves per item, per epoch. Format selection is path-name sniffing with a fall-through (`mace/tools/run_train_utils.py:32-162`): a directory named `*_lmdb` or containing `.lmdb`/`.aselmdb` files (`:80-91`), otherwise `.h5`/`.hdf5` files in it (`:93-106`), otherwise the substring `"lmdb"`/`"aselmdb"` anywhere in the lowercased path (`:108-117`), otherwise *try* HDF5 (`:119-130`); for a plain file, suffix dispatch (`:132-152`) and then the final fall-through "Attempting to load as LMDB" (`:154-162`) — a typo'd path is silently attempted as LMDB. There is no registry, no extension mechanism, and no error contract: legacy LMDB prints and **returns `None`** into the batch on read errors (`mace/data/lmdb_dataset.py:30-36`), and a second swallowed `except` around graph construction (`:59-68`) leaves `atomic_data` unbound so the failure resurfaces as an unrelated `UnboundLocalError`. Dataset statistics are computed in two divergent implementations — in-memory over a torch DataLoader (`mace/modules/utils.py:476-654`: `compute_mean_std_atomic_inter_energy:437`, `compute_mean_rms_energy_forces:482`, `compute_avg_num_neighbors:541`, `compute_statistics:554`, `compute_rms_dipoles:603`; E0 least-squares in `mace/data/utils.py:361`, `compute_average_E0s`) vs `pool_compute_stats` inside preprocessing (`mace/cli/preprocess_data.py:60-85`), which stores the results as **stringified Python dicts** (`preprocess_data.py:250-260`) parsed back with `ast.literal_eval` (`mace/cli/run_train.py:313-334`).

This ticket builds the v1 replacement in `mace_core.data`: a `DataBackend` Protocol (index / slice / iterate) that is the *only* thing the pipeline sees, an entry-point registry with explicit resolution (no fall-through, ever), the XYZ/ASE backend, and **one** statistics implementation with the legacy functions as named parity oracles. Backends are equal at the API; they yield `Configuration` (CORE-3 (#1557)'s framework-free type, `mace/data/utils.py:92-102`), **never** graph dicts — graph building (neighbors, one-hot, collation) happens once, above every backend, in DATA-3 (#1571)'s framework adapter. That layering is what makes backends framework-free (the same backend feeds torch and jax) and lets `mace_core` compute statistics without torch.

**Interface & constraints:**

1. **The `DataBackend` Protocol** (`mace_core.data.backend`; imports neither torch nor jax):

```python
@runtime_checkable
class DataBackend(Protocol):
    """Map-style, lazy, picklable (open handles dropped in __getstate__, reopened on
    demand — required for dataloader workers and DDP; legacy HDF5Dataset already does
    this at mace/data/hdf5_dataset.py:38-43)."""
    name: str                                    # registry name, e.g. "xyz", "hdf5", "lmdb"

    @classmethod
    def sniff(cls, source: str) -> bool: ...     # cheap: extension/magic-bytes; no full parse
    @classmethod
    def open(cls, source: str, *, key_spec: KeySpecification,
             head: str = "Default") -> "DataBackend": ...

    def __len__(self) -> int: ...                # exact, cheap after open
    def __getitem__(self, idx: int) -> Configuration: ...      # random access
    def iter_range(self, start: int = 0, stop: int | None = None,
                   step: int = 1) -> Iterator[Configuration]:  # sequential fast path
        ...                                      # default impl: getitem loop; shard formats override
    def statistics(self) -> "DatasetStatistics | None": ...
        # precomputed stats if the format carries them; None => caller computes
    def metadata(self) -> "DatasetManifest | None": ...
    # DatasetManifest: the typed description a persisted dataset carries beside its
    # data — schema_version: int (1), per-shard configuration counts (so an exact
    # __len__ never assumes a uniform group size), the KeySpecification used at
    # write time, the z-table, r_max, the E0 provenance, and an optional embedded
    # DatasetStatistics. Defined here with its JSON round-trip. The in-memory XYZ
    # backend returns None; the shard writer that emits dataset.json is DATA-2 (#1570).
```

- **Random access + exact `__len__` is the required core** — it is what `torch.utils.data` map-style datasets, deterministic train/valid splits, and `DistributedSampler` need. Unsized purely-iterable sources are out of v1; an `IterableDataBackend` can be registered later without touching this Protocol.
- **DDP sharding is a sampler concern**, never a backend method — a backend never knows about ranks. `iter_range` exists so shard formats can serve rank-contiguous reads efficiently when a sampler asks (wired in TRN-3 (#1577)/TRN-4 (#1578)).
- **Error contract:** `__getitem__` **raises** (`DataBackendError` carrying source + index). The legacy print-and-return-`None` and the second swallowed `except` are pinned-away defects, not behavior to preserve.
- `key_spec` is CORE-3 (#1557)'s `KeySpecification` (legacy origin `mace/data/utils.py:28-47`, with `update_keyspec_from_kwargs` at `:49` and the thirteen `DefaultKeys` members at `mace/tools/default_keys.py:6-19`); property-name mapping is resolved at `open()`, so a `Configuration`'s `properties` dict always uses convention names downstream.

2. **Registry** (`mace_core.data.registry`) — entry-point group **`mace.data_backends`**, mirroring the kernel-backend registry pattern (BKD-1 (#1560)): discovery without loading; import failures **recorded, not raised** (a broken optional backend must not break `import`); resolution raises.

```toml
[project.entry-points."mace.data_backends"]        # in mace-core's pyproject; third-party wheels identical
xyz = "mace_core.data.backends.xyz:XYZBackend"     # ase.io-readable files (in-memory)
```

```python
def available_backends() -> dict[str, "EntryPoint"]: ...
def open_dataset(source: str, *, format: str = "auto",
                 key_spec: KeySpecification, head: str = "Default") -> DataBackend:
    """format=<name>: load that backend or raise UnknownBackendError listing available ones.
    format="auto": run each registered backend's sniff() in declared priority order
    (xyz > hdf5 > hdf5-legacy > lmdb); ZERO matches or MORE THAN ONE match is a HARD
    ERROR naming the candidates — never a fall-through."""
```

One backend = one module + one entry-point line + the parametrized conformance harness. Adding a format requires zero repo edits.

3. **XYZ backend** (`mace_core.data.backends.xyz`): ase.io-readable files, in-memory; parsing delegates to CORE-3 (#1557)'s `read_configurations(path, key_spec)` so the P0-7-pinned behavior (property-key variants, weights, isolated-atom detection — legacy `load_from_xyz`, `mace/data/utils.py:239`) is inherited, not reimplemented. `sniff` is cheap (extension check, no full parse).

4. **One statistics implementation** — `mace_core.data.statistics.compute_statistics(backend, z_table, r_max, atomic_energies) -> DatasetStatistics`. The isolated-atom energies are **passed in already resolved**, so the interaction-energy mean/std are always computed against the same E0 values that become the model buffer, whatever their origin (isolated-atom configs, a file, a foundation model via `estimate_e0s_from_foundation`, `mace/data/utils.py:391`, or a least-squares fit). Type them `dict[int, float]` for the single-head case and `dict[str, dict[int, float]]` when heads are present, and expose the least-squares fit as a separate function a caller may use to produce them (numpy; neighbor counts via `mace_core.neighbors` — the edge-set contract; the P0-7 brute-force minimum-image helper is the oracle implementation of that contract until DATA-3 (#1571)'s production module lands). It is used by *both* the future `mace data prepare` (CLI-1 (#1579)) and the in-memory path, so "parity vs legacy" is a single golden instead of two implementations to reconcile. Named legacy oracles: `mace/modules/utils.py:476,521,580,593,642` and `mace/data/utils.py:361`.

5. **`DatasetStatistics`** (dataclass; JSON round-trip with real types — no `str(dict)` / `ast.literal_eval`): `atomic_energies: dict[int, float]` (or per-head), `avg_num_neighbors: float`, `mean: float`, `std: float`, `atomic_numbers: list[int]`, `r_max: float`. The legacy consumer this replaces reads exactly these five keys back out of `statistics.json` (`mace/cli/run_train.py:319-334`), so the field set is a compatibility surface, not a fresh design.

6. **Conformance harness** (`data_backend_conformance(backend)`, parametrized): len / getitem / pickle round-trip / raises-on-corruption / statistics parity on a tiny fixture. DATA-2 (#1570)'s HDF5/LMDB backends and any third-party plugin get tested for free by pointing the harness at them.

**Task:**
1. `mace_core.data.backend`: the `DataBackend` Protocol, `DataBackendError`, `DatasetStatistics` + `DatasetManifest` dataclasses with typed-JSON round-trip — no torch/jax imports.
2. `mace_core.data.registry`: `available_backends()` + `open_dataset()` with explicit-name and `"auto"` sniff resolution, declared priority order, hard error naming candidates on zero/multiple matches; discovery records import failures without raising.
3. `mace_core.data.backends.xyz`: the XYZ/ASE backend over CORE-3 (#1557)'s `read_configurations`, registered under `mace.data_backends`.
4. `mace_core.data.statistics.compute_statistics`: numpy implementation of avg-num-neighbors, E0 least-squares, interaction-energy mean/std, force RMS and dipole RMS per the legacy oracle set, consuming any `DataBackend`.
5. Parity tests: statistics values vs the named legacy oracles on a shared fixture; assert the fall-through and return-`None` anti-patterns are structurally impossible (typo'd path → hard error; corrupt item → `DataBackendError`).
6. The parametrized `data_backend_conformance` harness, green for `xyz`; a dummy backend registered via an entry point in a test package is discovered and opened.

**Out of scope:** HDF5 v2 / `hdf5-legacy` / LMDB backends and the shard writer + `dataset.json` manifest (DATA-2 (#1570)); the `mace data prepare` CLI surface (CLI-1 (#1579)); graph construction, neighbor-list module, collation, and the torch `GraphDataset` adapter (DATA-3 (#1571)); DDP sampler wiring (TRN-3 (#1577)/TRN-4 (#1578)).

**Acceptance criteria:**
- [ ] `mace_core.data` imports neither torch nor jax (asserted by a test).
- [ ] `open_dataset(format="auto")` on a path matching zero or multiple backends raises a hard error naming the candidates; an unknown explicit `format=` raises `UnknownBackendError` listing available backends — no fall-through path exists.
- [ ] XYZ backend: `sniff`/`open`/`__len__`/`__getitem__`/`iter_range` conform; `Configuration.properties` uses convention names for every property-key variant covered by the CORE-3 (#1557)/P0-7 parsing cases.
- [ ] `__getitem__` on a corrupt/unreadable item raises `DataBackendError` with source + index; a backend instance pickle round-trips and reads identically after unpickling.
- [ ] `compute_statistics` matches the legacy oracles (`compute_avg_num_neighbors`, `compute_average_E0s` least-squares, interaction-energy mean/std, force RMS, dipole RMS) on the shared fixture — the statistics-parity golden.
- [ ] `DatasetStatistics` round-trips through JSON with real types (ints as ints, dict keys as atomic numbers), no `ast.literal_eval` anywhere, and carries the five keys legacy's `statistics.json` consumer reads.

**Inventory gaps assigned here:**
- [ ] `--mean` / `--std` (`mace/tools/arg_parser.py:510,517`) as statistics overrides (pairs with the P0-6 case) — untested today.

**Verify:**
```bash
python -m pytest packages/mace-core/tests/data/test_backend_protocol.py packages/mace-core/tests/data/test_registry.py -v
python -m pytest packages/mace-core/tests/data/test_xyz_backend.py -v                 # conformance harness on xyz
python -m pytest packages/mace-core/tests/data/test_statistics_parity.py -v          # vs named legacy oracles
```
**Review focus:** the statistics-parity golden — these values (E0s, mean/std, avg neighbors) seed model scale/shift, so a silent deviation shifts every model trained on v1; and the Protocol surface itself, which DATA-2 (#1570), TRN-3 (#1577), and third-party plugins build against and which cannot cheaply change after Phase 2.

---
